### PR TITLE
API: ExtensionArray.argsort places the missing value at the end 

### PR DESCRIPTION
**Pull request by makbigc**
continue #26354 

1. Categorical will be handled in a separate PR.
2. benchmark result:
```
       before           after         ratio
     [989f912e]       [dc0ef701]
     <master>         <enh-21801-2>
+     11.4±0.03μs       36.8±0.1μs     3.23  algorithms.EASorting.time_argsort(1000)
+       785±0.5μs      1.26±0.03ms     1.60  algorithms.EASorting.time_argsort(100000)

```

xref https://github.com/pandas-dev/pandas/issues/21801 (doesn't close; Need to update existing EAs)

### CODE REVIEW

File: asv_bench/benchmarks/algorithms.py [hunk 1]
```diff
@@ -141,4 +141,16 @@ def time_quantile(self, quantile, interpolation, dtype):
         self.idx.quantile(quantile, interpolation=interpolation)
 
 
+class EASorting:
```
Comment (TomAugspurger) [inline asv_bench/benchmarks/algorithms.py]: Can you make this name of this class specific to IntegerNA?

File: doc/source/whatsnew/v0.25.0.rst [hunk 2]
```diff
@@ -562,6 +562,7 @@ Other API changes
 - Removed support of gtk package for clipboards (:issue:`26563`)
 - Using an unsupported version of Beautiful Soup 4 will now raise an ``ImportError`` instead of a ``ValueError`` (:issue:`27063`)
 - :meth:`Series.to_excel` and :meth:`DataFrame.to_excel` will now raise a ``ValueError`` when saving timezone aware data. (:issue:`27008`, :issue:`7056`)
+- :meth:`ExtensionArray.argsort` places NA values at the end of the sorted array, except :class:`Categorical` (:issue:`21801`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.25.0.rst]: Can remove the "except Categorical". I'll make that change before the release.

File: pandas/tests/extension/test_categorical.py [hunk 3]
```diff
@@ -193,6 +193,11 @@ def test_searchsorted(self, data_for_sorting):
         if not data_for_sorting.ordered:
             raise pytest.skip(reason="searchsorted requires ordered data.")
 
+    def test_argsort_nan_last(self, data_missing_for_sorting):
```
Comment (TomAugspurger) [inline pandas/tests/extension/test_categorical.py]: You can just remove this here.

**Review by TomAugspurger — COMMENTED** [review]
@makbigc can you add a test in the base extension array tests? 

```diff
diff --git a/pandas/tests/extension/base/methods.py b/pandas/tests/extension/base/methods.py
index d9e61e6a2..17315a52e 100644
--- a/pandas/tests/extension/base/methods.py
+++ b/pandas/tests/extension/base/methods.py
@@ -52,6 +52,11 @@ class BaseMethodsTests(BaseExtensionTests):
         expected = pd.Series(np.array([1, -1, 0], dtype=np.int64))
         self.assert_series_equal(result, expected)
 
+    def test_argsort_missing_array(self, data_missing_for_sorting):
+        result = np.argsort(data_missing_for_sorting)
+        expected = data_missing_for_sorting.take([2, 0, 1])
+        self.assert_extension_array_equal(result, expected)
+
     @pytest.mark.parametrize('na_position, expected', [
         ('last', np.array([2, 0, 1], dtype=np.dtype('intp'))),
         ('first', np.array([1, 2, 0], dtype=np.dtype('intp')))
```

If you don't have time today, LMK. I may push some changes to get this in.

File: pandas/core/arrays/categorical.py [hunk 4]
```diff
@@ -1531,7 +1531,7 @@ def check_for_ordered(self, op):
     def _values_for_argsort(self):
         return self._codes.copy()
 
-    def argsort(self, *args, **kwargs):
+    def argsort(self, ascending=True, kind='quicksort', *args, **kwargs):
         # TODO(PY2): use correct signature
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: can remove this comment? is the doc-string updated?

Comment (TomAugspurger) [reply pandas/core/arrays/categorical.py]: Yeah. I was going to do Categorical as a followup, but it seems tiny so I'll do it here.

**Comment by TomAugspurger** [comment]
FYI @jorisvandenbossche just doing the breaking Categorical change here.

**Comment by TomAugspurger** [comment]
As expected, a few failures in our test suite. Going through them next.

I think https://travis-ci.org/pandas-dev/pandas/jobs/553767675#L2498 is buggy / defined strangely.  (DataFrame.sort_index with a CategoricalIndex and ascending=False)



```python
# 0.24
In [2]: df = pd.DataFrame({"A": list(range(6))}, index=pd.Categorical(['a', 'a', 'b', 'b', 'c', 'a'], categories=['c', 'a', 'b']))

In [3]: df.sort_index(ascending=False)
Out[3]:
   A
b  3
b  2
a  5
a  1
a  0
c  4

In [4]: df.set_index(df.index.codes).sort_index(ascending=False)
Out[4]:
   A
2  2
2  3
1  0
1  1
1  5
0  4

```

```python
# this PR

In [3]: df.sort_index()
Out[3]:
   A
c  4
a  0
a  1
a  5
b  2
b  3

In [4]: df.set_index(df.index.codes).sort_index()
Out[4]:
   A
0  4
1  0
1  1
1  5
2  2
2  3
```

I think that sorting by a CategoricalIndex should be the same as sorting by its codes. 0.24 (and master, before the PR)  doesn't have that property.

**Comment by jorisvandenbossche** [comment]
It's probably more a comment on the changes in https://github.com/pandas-dev/pandas/pull/26854, but it seems that `nargsort` is now always calling `_values_for_argsort` for EAs. But, in the comments of `ExtensionArray.argsort`, we say that there are two ways to override argsort: have `_values_for_argsort`, or implement `argsort` yourself for full control. 

So the current indirection seems a bit wrong. At least depending on where `nargsort` is called internally, it should call the `argsort` method of the EA, to let the EA have full control over it.

**Comment by TomAugspurger** [comment]
Hmm, so we would want to tell users that they should *always* override `_values_for_argsort`, since just doing `.argsort` may not be enough. OK, I'll try to write something about that.

**Comment by jorisvandenbossche** [comment]
> Hmm, so we would want to tell users that they should always override _values_for_argsort, since just doing .argsort may not be enough.

Not fully sure. We *could* also pass `_values_for_argsort` to `nargsort` from withing EA.argsort, and let `nargsort` call `EA.argsort` (I think, might be missing something in this indirection web).

**Comment by TomAugspurger** [comment]
I'm also having trouble with the indirection.

AFAICT, with this PR we end up in a situation where `ExtensionArray.argsort` could differ from other places in pandas that do `nargsort(ExtensionArray)`. Since `nargsort` Does pandas' usual sorting on the `ExtensionArray._values_for_argsort`...

**Comment by TomAugspurger** [comment]
All green. I had to slightly adjust the test in https://github.com/pandas-dev/pandas/pull/27137/commits/47c6b57bb7d2b7959f5f96db6aeebe803d2dc1b0

I don't have a 32-bit machine to test on, but it seems like we may be inconsistent with whether `_values_for_argsort` returns an int32 or int64 dtype array. It doesn't seem to matter in the algorithm, so I cast to int64 before making the assert.

**Comment by jreback** [comment]
> All green. I had to slightly adjust the test in [47c6b57](https://github.com/pandas-dev/pandas/commit/47c6b57bb7d2b7959f5f96db6aeebe803d2dc1b0)
> 
> I don't have a 32-bit machine to test on, but it seems like we may be inconsistent with whether `_values_for_argsort` returns an int32 or int64 dtype array. It doesn't seem to matter in the algorithm, so I cast to int64 before making the assert.

I believe this will break on the 32-bit builds; argsort is supposed to return platform ints.

So I guess our current interface is kind wrong. But let's merge this and consider that.

File: pandas/core/arrays/base.py [hunk 5]
```diff
@@ -409,7 +410,8 @@ def argsort(self, ascending=True, kind='quicksort', *args, **kwargs):
         Returns
         -------
         index_array : ndarray
```
Comment (jreback) [inline pandas/core/arrays/base.py]: so we aren't specific that these are actually intp (and NOT int64's)

**Comment by jorisvandenbossche** [comment]
> AFAICT, with this PR we end up in a situation where ExtensionArray.argsort could differ from other places in pandas that do nargsort(ExtensionArray). Since nargsort Does pandas' usual sorting on the ExtensionArray._values_for_argsort...

Let's open an issue about that to fix after the RC before final 0.25.0 ?

**Comment by TomAugspurger** [comment]
Writing one now.

**Comment by TomAugspurger** [comment]
https://github.com/pandas-dev/pandas/issues/27218, but my thinking on this
may be a bit unclear.

**Pull request merged**